Fix review dismissals bug

devinburnette opened the pull request:
Closes https://github.com/palantir/policy-bot/issues/480

In https://github.com/palantir/policy-bot/pull/463 we introduced a feature to dismiss stale reviews, while iterating on it some additional logic slipped in to also dismiss reviews that didn't match review patterns if people were using the `github_review_comment_patterns` field in their policies. This worked well until a predicate was included in a policy causing certain reviews to be skipped from evaluation and therefore not added to the list of approved candidates. This manifested in buggy/broken behavior with a misleading dismissal reason until the predicate was satisfied.

After deliberating in the issue above, I think we agreed that the dismissal of non-matching patterns is just a little too heavy handed and doesn't really fit well with dismissing "stale" reviews. In other words, if we aren't also dismissing all other reviews that can never satisfy a rule under any condition than the feature seems broken. So this PR removes dismissal of non-matching patterns, but keeps dismissal for both edited and invalidated reviews, classifying those as "stale" if people have the `ignore_edited_comments` or `invalidate_on_push` fields set to true in their policies. In addition, this PR also highlights which patterns will satisfy which rules on the details page in hopes to clear up confusion for folks using approval comments.

![Screen Shot 2022-10-13 at 6 21 30 PM](https://user-images.githubusercontent.com/13012689/195724431-b93b263b-da65-4f78-912c-b9002a93f78e.png)

devinburnette committed:
only dismiss edited or invalidated reviews

devinburnette committed:
add patterns to details page

devinburnette commented on policy/approval/approve.go:
```diff
@@ -111,7 +111,8 @@ func (r *Rule) Evaluate(ctx context.Context, prctx pull.Context) (res common.Res
 	res.Description = r.Description
 	res.Status = common.StatusSkipped
-	res.Requires = r.Requires.Actors
+	res.Requires = common.Requires{Count: r.Requires.Count, Actors: r.Requires.Actors}
```
this felt a little funny, but i kept getting cyclical dependency issues when just trying to use the existing `approval.Requires`

bluekeyes replied:
Moving types to `common` is normal as they end up used in `common.Result`, so I think this is fine.

devinburnette commented on server/handler/frontend.go:
```diff
@@ -60,14 +60,20 @@ func LoadTemplates(c *FilesConfig, basePath string, githubURL string) (templatet
-		"getRequires": func(results *common.Result) map[string][]Membership {
-			return getRequires(results, strings.TrimSuffix(githubURL, "/"))
+		"hasPatterns": func(results *common.Result) bool {
+			return !results.Methods.IsEmpty()
```
there's one thing I think I'm missing here and that's what if a github review approval (without comments) would satisfy the policy just fine? I think i can make it only show the Patterns if `github_review: true` and `github_review_comment_patterns` is not set.

devinburnette replied:
but then it's like the logic is slipping into the frontend a little bit.

bluekeyes replied:
What if instead of thinking of this as patterns-only, we show all methods? I think that removes the need for `hasPatterns` (every rule will have methods) and then we can change `getPatterns` to `getMethods` and update the template accordingly (see later comments)

devinburnette replied:
sgtm

devinburnette commented on policy/common/methods.go:
```diff
@@ -179,3 +179,7 @@ func (m *Methods) BodyMatches(prBody string) bool {
 }
+
+func (m *Methods) IsEmpty() bool {
+	return m == nil || (len(m.Comments) == 0 && len(m.CommentPatterns) == 0 && len(m.GithubReviewCommentPatterns) == 0 && len(m.BodyPatterns) == 0)
```
oh i can just put it here, and maybe call this something different.

bluekeyes replied:
I think this also needs to check that `m.GithubReview` is non-nil to account for all fields.

devinburnette replied:
yea i realized that, but with the new approach, i think we don't need it anymore.

bluekeyes commented on policy/common/result.go:
```diff
@@ -59,14 +59,20 @@ type ReviewRequestRule struct {
 	Mode RequestMode
 }
 
+type Requires struct {
```
Instead of duplicating `Requires` in this package, I think it might be better to move `approval.Requires` to this package and then fix the references in the `approval` packages. That should reduce confusion between the two types in the future since they're otherwise the same.

bluekeyes commented on server/handler/frontend.go:
```diff
@@ -60,14 +60,20 @@ func LoadTemplates(c *FilesConfig, basePath string, githubURL string) (templatet
 		},
-		"hasRequires": func(requires common.Actors) bool {
-			return len(requires.Users) > 0 || len(requires.Teams) > 0 || len(requires.Organizations) > 0
+		"hasActors": func(requires *common.Requires) bool {
```
Thanks for renaming some of these existing functions, on review the old naming scheme was kind of confusing.

bluekeyes commented on server/templates/details.html.tmpl:
```diff
@@ -65,14 +65,19 @@
+        {{if (hasPatterns .)}}
           <div class="pt-2">
-          {{template "result-requires-permissions-details" .}}
+          {{template "result-patterns-details" .}}
```
I think it would make sense to reorder these templates so it reads something like:

> This rule needs N approvals from people in these groups or people with these permissions. Approvals can use these patterns.

As currently structured, it kind of implies that the sections are independent so that the pattern section doesn't also require the approvers have certain team membership/permissions.

bluekeyes commented on server/templates/details.html.tmpl:
```diff
@@ -142,10 +147,22 @@
-{{define "result-requires-details"}}
-  <b class="font-bold text-sm">This rule requires review from</b>
+{{define "result-patterns-details"}}
+  <b class="font-bold text-sm">This rule requires {{.Requires.Count}} review(s) using any of these patterns</b>
```
Good call adding the count to all of these - I'm not sure why we didn't add it earlier. I think it would be better to only show the count once before all the other approval details. With the count repeated, people could interpret it as if they need N reviews for each section instead of N reviews total for the rule.

devinburnette replied:
yea that's fair, i was trying to play with the wording so that wouldn't be the case, but I think this is a better idea.

bluekeyes reviewed (commented):
Thanks for fixing this and expanding the details view! I had some suggestions on the UI part, but the rest of the change looks good.

devinburnette committed:
implement feedback

devinburnette commented on server/handler/frontend.go:
```diff
@@ -91,24 +94,47 @@ func Static(prefix string, c *FilesConfig) http.Handler {
+			}
+		} else {
+			patternInfo["Github Review State"] = append(patternInfo["Github Review State"], string(result.Methods.GithubReviewState))
+		}
```
this is the only part where i feel like the logic slipped into the view a little, but i thought the result would be more helpful for humans to read.

bluekeyes committed:
Restore custom Requires type in disapproval

Disapproval does not support counts, so using the common type is
misleading. Add an new comment explaining this.

bluekeyes committed:
Add nil check

This shouldn't be nil in practice, but would prefer to not panic
over this if something changes in the future.

bluekeyes committed:
Fix typo in field name

bluekeyes reviewed (approved):
Thanks for addressing the comments and sorry about the delayed review, this looks good to merge
